django/db/models/query.py

```diff
@@ -464,7 +464,41 @@ def _prepare_for_bulk_create(self, objs):
                 obj.pk = obj._meta.pk.get_pk_value_on_save(obj)
             obj._prepare_related_fields_for_save(operation_name='bulk_create')
 
-    def bulk_create(self, objs, batch_size=None, ignore_conflicts=False):
+    def _select_on_conflict(
+        self, ignore_conflicts, update_conflicts, update_fields, unique_fields
+    ):
+        if ignore_conflicts and update_conflicts:
+            raise ValueError(
+                'You can only assign one conflicts plan, ignore_conflicts or update_conflicts'
+            )
+        if ignore_conflicts:
+            self._check_on_conflicts_supported(OnConflict.Ignore, unique_fields)
+            result = OnConflict.Ignore
+        elif update_conflicts:
+            self._check_on_conflicts_supported(OnConflict.Update, unique_fields)
+            if not update_fields:
+                raise IntegrityError(
+                    'You need to specify which fields you want to update'
+                )
+            for f in update_fields:
+                try:
+                    if f == 'pk':
+                        f = self.model._meta.pk.name
+                    self.model._meta.get_field(f)
+                except exceptions.FieldDoesNotExist:
+                    raise IntegrityError(
+                        'Field you specify is not in table: %s' % f
+                    )
+            result = OnConflict.Update
+        else:
+            result = None
+
+        return result
```

Comment:
Avoid using `IntegrityError` here. It is a `ValueError` to not provide `update_fields` with `update_conflicts`.

```suggestion
        if ignore_conflicts and update_conflicts:
            raise ValueError('The ignore_conflicts and update_conflicts flags are mutually exclusive.')
        if ignore_conflicts:
            self._check_on_conflicts_supported(OnConflict.IGNORE, unique_fields)
            return OnConflict.IGNORE
        if update_conflicts:
            self._check_on_conflicts_supported(OnConflict.UPDATE, unique_fields)
            if not update_fields:
                raise ValueError('The update_conflicts flag requires update_fields to be specified.')
            for name in update_fields:
                if name == 'pk':
                    name = self.model._meta.pk.name
                try:
                    self.model._meta.get_field(name)
                except exceptions.FieldDoesNotExist:
                    raise ValueError(f'The update_fields list contains an unknown field: {name}.')
            return OnConflict.UPDATE
        return None
```